### DISCUSSION TITLE
Unable to use nerdctl in rootless mode: iptables not found in $PATH despite installed

### DISCUSSION THREAD

**Discussion by sarunint** [containerd/nerdctl#1536] [Q&A] [answered]
Hi,

I somehow cannot run `nerdctl` in rootless mode. The error I got is:

```
sarunint@laptop:~> nerdctl run -it --rm alpine
docker.io/library/alpine:latest:                                                  resolved       |++++++++++++++++++++++++++++++++++++++| 
index-sha256:b95359c2505145f16c6aa384f9cc74eeff78eb36d308ca4fd902eeeb0a0b161b:    done           |++++++++++++++++++++++++++++++++++++++| 
manifest-sha256:3d426b0bfc361d6e8303f51459f17782b219dece42a1c7fe463b6014b189c86d: done           |++++++++++++++++++++++++++++++++++++++| 
config-sha256:bfe296a525011f7eb76075d688c681ca4feaad5afe3b142b36e30f1a171dc99a:   done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:ca7dd9ec2225f2385955c43b2379305acd51543c28cf1d4e94522b3d94cce3ce:    done           |++++++++++++++++++++++++++++++++++++++| 
elapsed: 9.5 s                                                                    total:  2.7 Mi (288.8 KiB/s)                                     
FATA[0009] failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error running hook #0: error running hook: exit status 1, stdout: , stderr: time="2022-11-22T19:38:47+07:00" level=fatal msg="failed to call cni.Setup: plugin type=\"bridge\" failed (add): failed to locate iptables: exec: \"iptables\": executable file not found in $PATH"
Failed to write to log, write /home/sarunint/.local/share/nerdctl/1935db59/containers/default/2fa383a19e1cf99c4042d37920f48a423962d36a67df78b2491fb4c8d6aa612b/oci-hook.createRuntime.log: file already closed: unknown
```

The error mentioned `\"iptables\": executable file not found in $PATH"`, so I checked my system. I have `iptables` installed, the full-path is `/usr/sbin/iptables`.

My system is running openSUSE Tumbleweed. `nerdctl` is installed from official openSUSE repository.

Any help would be appreciated!

Thanks.

**Comment by Zheaoli** [comment]
Which OpenSUSE version you used?

**Reply by sarunint** [reply]
openSUSE Tumbleweed, it is a rolling release distribution

**Reply by Zheaoli** [reply]
Have you ever setup the rootless mode by using script https://github.com/containerd/nerdctl/blob/main/extras/rootless/containerd-rootless.sh?

**Reply by sarunint** [reply]
I ran using the `containerd-rootless-setuptool.sh install`. Is that enough?

**Comment by LordGnomish** [comment]
Hi!

! I have the same issue here, did you solve this problem?

Best,

**Comment by LordGnomish** [comment]
I also use ```containerd-rootless-setuptool.sh install```

**Comment by pedroscaff** [comment]
I had the same issue and in my system (debian bullseye) `iptables` and `ip6tables` was installed only under `/usr/sbin`. A quick fix was to make a link in `/usr/local/bin`:

- `ln -s /usr/sbin/iptables /usr/local/bin/iptables`
- `ln -s /usr/sbin/ip6tables /usr/local/bin/ip6tables`

**Reply by davidecavestro** [reply]
Same for Debian 12/Bookworm

**Reply by midnight-wonderer** [reply]
For those who want to stay true to the rootless spirit:
~~~bash
ln -s /usr/sbin/iptables ~/.local/bin/iptables
ln -s /usr/sbin/ip6tables ~/.local/bin/ip6tables
~~~

To the original asker (@sarunint)
- I am on Tumbleweed, too; the symlinks above work and require no root.
- My name is Sarun, too.

**Comment by zgfh** [comment]
this work for me （debain 12.2）
```
apt install iptables
```

**Comment by jeffrson** [comment]
So is this a bug in `containerd-rootless-setuptool.sh install`? Should it create links in ~/.local/bin (given that profile start scripts append that to the PATH) - and remove those in `uninstall`?

Currently it still seems, rootless setup as documented does not work for most new installations since `iptables` is located in `/usr/sbin` (if at all). What is the correct solution?

**Reply by AkihiroSuda** [reply]
> given that profile start scripts append that to the PATH

What do you mean?

> does not work for most new installations

What's your distro?
The script has been tested to work on Debian, Ubuntu, Fedora, openSUSE, and Arch Linux.

**Reply by jeffrson** [reply]
I'm very sorry that I wasn't clear enough :-/

I'm using Debian. The script itself runs fine actually - however it doesn't really enable rootless mode, because `nerdctl run` does not work later on. The error message is the same as written in the question. 'iptables' cannot be found. IMO, it's a bug that it does not work after `containerd-rootless-setuptool.sh install`.

I wonder why it works for you. Where do you have `iptables` and `ip6tables`? Yes, for Debian, that iptables package has to be installed, but that's not the point.

AFAICT, nerdctl is using go-iptables, and it seems, that this relies on finding `iptables` in $path. In order to achieve that, one of the other comments recommended to link .local/bin/iptables to /usr/sbin/iptables (similarly with ip6tables). This is something that rootless setup script could do. Of course it requires .local/bin appended (or prepended) to the users $path. Which is generally done in .bashrc (*). This is what I meant with the other remark.

Another possible solution could be that nerdctl tries to find iptables in /usr/sbin and hand this path over to go-iptables via options.

(*) it may be necessary to source .bashrc or to relogin, because .local/bin might not be added to $path, if it doesn't exist during logon.

**Reply by AkihiroSuda** [reply]
Works fine for me (`limactl start template://debian`)
```console
$ which iptables
/usr/sbin/iptables

$ echo $PATH
/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games:/usr/sbin:/sbin:/usr/sbin:/sbin

$ nerdctl info
Client:
 Namespace:	default
 Debug Mode:	false

Server:
 Server Version: v2.0.5
 Storage Driver: overlayfs
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Log:     fluentd journald json-file none syslog
  Storage: native overlayfs fuse-overlayfs stargz
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
  rootless
 Kernel Version:   6.1.0-34-cloud-arm64
 Operating System: Debian GNU/Linux 12 (bookworm)
 OSType:           linux
 Architecture:     aarch64
 CPUs:             4
 Total Memory:     3.836GiB
 Name:             lima-debian
 ID:               cec972f2-401d-462e-a29f-875bf4be2894

WARNING: AppArmor profile "nerdctl-default" is not loaded.
         Use 'sudo nerdctl apparmor load' if you prefer to use AppArmor with rootless mode.
         This warning is negligible if you do not intend to use AppArmor.
WARNING: IPv4 forwarding is disabled

$ nerdctl run -d --name nginx -p 80:80 nginx:alpine
1fbf8a84031dfa218acda88c1323be6bfdc53bd03785f7f647b13873e61d0f5e

$ curl localhost
<!DOCTYPE html>
<html>
<head>
<title>Welcome to nginx!</title>
<style>
html { color-scheme: light dark; }
body { width: 35em; margin: 0 auto;
font-family: Tahoma, Verdana, Arial, sans-serif; }
</style>
</head>
<body>
<h1>Welcome to nginx!</h1>
<p>If you see this page, the nginx web server is successfully installed and
working. Further configuration is required.</p>

<p>For online documentation and support please refer to
<a href="http://nginx.org/">nginx.org</a>.<br/>
Commercial support is available at
<a href="http://nginx.com/">nginx.com</a>.</p>

<p><em>Thank you for using nginx.</em></p>
</body>
</html>
```

**Reply by AkihiroSuda** [reply]
FYI: I have the following lines in `~/.profile`, `~/.bash_profile`, and `.zshrc` (via https://github.com/lima-vm/lima/blob/v1.0.7/pkg/cidata/cidata.TEMPLATE.d/boot/20-rootless-base.sh#L28-L30)
```bash
PATH="$PATH:/usr/sbin:/sbin"
export PATH
```

**Reply by AkihiroSuda** [reply]
Does this work for you?
```diff
diff --git a/pkg/cmd/container/create.go b/pkg/cmd/container/create.go
index 2b3b885d..39b386a5 100644
--- a/pkg/cmd/container/create.go
+++ b/pkg/cmd/container/create.go
@@ -540,6 +540,8 @@ func withNerdctlOCIHook(cmd string, args []string) (oci.SpecOpts, error) {
        }
 
        args = append([]string{cmd}, append(args, "internal", "oci-hook")...)
+       // sbin is appended for iptables https://github.com/containerd/nerdctl/discussions/1536
+       env := append([]string{"PATH=" + os.Getenv("PATH") + "/usr/sbin:/sbin"}, os.Environ()...)
        return func(_ context.Context, _ oci.Client, _ *containers.Container, s *specs.Spec) error {
                if s.Hooks == nil {
                        s.Hooks = &specs.Hooks{}
@@ -548,14 +550,14 @@ func withNerdctlOCIHook(cmd string, args []string) (oci.SpecOpts, error) {
                s.Hooks.CreateRuntime = append(s.Hooks.CreateRuntime, specs.Hook{
                        Path: cmd,
                        Args: crArgs,
-                       Env:  os.Environ(),
+                       Env:  env,
                })
                argsCopy := append([]string(nil), args...)
                psArgs := append(argsCopy, "postStop")
                s.Hooks.Poststop = append(s.Hooks.Poststop, specs.Hook{
                        Path: cmd,
                        Args: psArgs,
-                       Env:  os.Environ(),
+                       Env:  env,
                })
                return nil
        }, nil
```

**Reply by jeffrson** [reply]
Yes, indeed, if $PATH contains {/usr}/sbin it will work. For me, Debian netinstall, it's `/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games`, however.

The patch looks fine, but I haven't tried it yet. Will report if done.

**Reply by jeffrson** [reply]
Well, didn't work, unfortunately. First, it's missing a ':' before `/usr/sbin`. Then it seems, that `os.Environ()` overrides those changes.

I'm not a Go developer (yet), so this may not be quite correct, but it works for me anyway :-)

```
+       env := append(os.Environ(), "PATH=" + os.Getenv("PATH") + ":/usr/sbin:/sbin")
```

Thank you for looking into it!

**Reply by jeffrson** [reply]
It seems like this (simpler) patch does it as well:

```diff
diff --git a/pkg/cmd/container/create.go b/pkg/cmd/container/create.go
index c727dce0..43579a91 100644
--- a/pkg/cmd/container/create.go
+++ b/pkg/cmd/container/create.go
@@ -568,6 +568,8 @@ func withNerdctlOCIHook(cmd string, args []string) (oci.SpecOpts, error) {
        }

        args = append([]string{cmd}, append(args, "internal", "oci-hook")...)
+       // sbin is appended for iptables https://github.com/containerd/nerdctl/discussions/1536
+       os.Setenv("PATH", os.Getenv("PATH") + ":/usr/sbin:/sbin")
        return func(_ context.Context, _ oci.Client, _ *containers.Container, s *specs.Spec) error {
                if s.Hooks == nil {
                        s.Hooks = &specs.Hooks{}
```

**Reply by AkihiroSuda** [reply, marked as answer]
Thanks for testing, opened a PR (v2.1.0) to fix the issue
- https://github.com/containerd/nerdctl/pull/4208

**Reply by AkihiroSuda** [reply]
I avoided `os.Setenv` as I did't like altering the env in the halfway of the `nerdctl create` steps
